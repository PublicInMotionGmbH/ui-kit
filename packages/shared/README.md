# Talixo Shared

Shared data between whole UI Kit.

## How to install

Package is available as `@talixo/shared` in NPM registry, so you can use it in your project
using `npm install @talixo/shared --save` or `yarn add @talixo/shared`.

## Requirements

Your package should additionally have some extra dependencies:

- `classnames: ^2.2.5`

These packages are required by `@talixo/shared`, but you have to install them manually,
to avoid having different versions of these in your application.
