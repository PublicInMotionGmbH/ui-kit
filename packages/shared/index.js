import applyAnyClassNameModifiers from './src/applyAnyClassNameModifiers'
import applyClassNameModifiers from './src/applyClassNameModifiers'
import buildClassName from './src/buildClassName'
import prefix from './src/prefix'

export {
  applyAnyClassNameModifiers,
  applyClassNameModifiers,
  buildClassName,
  prefix
}
