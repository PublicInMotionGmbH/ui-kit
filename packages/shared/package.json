{
  "name": "@talixo/shared",
  "version": "1.0.0",
  "description": "Talixo common configuration",
  "author": "Public in Motion GmbH",
  "main": "index.js",
  "module": "index.js",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/PublicInMotionGmbH/ui-kit.git"
  },
  "bugs": {
    "url": "https://github.com/PublicInMotionGmbH/ui-kit/issues"
  },
  "devDependencies": {
    "rollup": "^0.57.1"
  },
  "dependencies": {
    "normalize.css": "^8.0.0"
  }
}
