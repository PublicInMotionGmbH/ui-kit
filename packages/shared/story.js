import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'

import createController from './story/createController'

/**
 * Build styles for story
 *
 * @param {object} stylesheet
 * @param {object} options
 * @returns {object}
 */
const styles = (stylesheet, options) => ({
  ...stylesheet,
  infoBody: {
    ...stylesheet.infoBody,
    boxShadow: 'none',
    border: 0,
    padding: 0
  },
  propTableHead: options.propTables && options.propTables.length > 1 ? {

  } : {
    display: 'none'
  },
  source: {
    ...stylesheet.source,
    h1: {
      ...stylesheet.source.h1,
      marginBottom: '15px'
    }
  }
})

/**
 * Create factory for stories
 *
 * @param {string} name
 * @param {object} module
 * @param {object} defaultOptions
 *
 * @returns {function(name: string, description: string, render: function)|function(name: string, description: string, render: function, infoOptions: object)|{ controlled: function(name: string, description: string, render: function, getInitialState: function, infoOptions: object)|function(name: string, description: string, render: function, getInitialState: function, infoOptions: object) }}
 */
export function createStoriesFactory (name, module, defaultOptions = {}) {
  const stories = storiesOf(name, module)

  function addStory (name, description, render, infoOptions = {}) {
    // When rendering function expect any arguments,
    // It means that it's controlled
    if (render.length > 0) {
      render = createController(render)

      infoOptions = {
        ...infoOptions,
        propTablesExclude: [
          render.Controller,
          ...(infoOptions.propTablesExclude || [])
        ]
      }
    }

    // Build options for story
    const options = {
      header: true,
      inline: true,
      text: description,
      ...defaultOptions,
      ...infoOptions
    }

    // Build styles factory
    const stylesFunction = options.styles
      ? stylesheet => options.styles(styles(stylesheet, options), options)
      : stylesheet => styles(stylesheet, options)

    // Create story
    return stories.add(name, withInfo({
      ...options,
      styles: stylesFunction
    })(render))
  }

  return addStory
}

/**
 * Get description from README file
 *
 * @param {string} content
 * @returns {string}
 */
export function getReadmeDescription (content) {
  return content
    .replace(/<p><h1([^>]*)>(.*)<\/h1><\/p>/g, '<h1$1>$2</h1>')
    .replace(/^<h1[^>]*>.*<\/h1>/, '')
    .replace(/(<p>)?<h[1-6][> ][^]*$/i, '')
}
