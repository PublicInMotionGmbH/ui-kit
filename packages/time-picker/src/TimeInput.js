import React from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'

import { buildClassName } from '@talixo/shared'
import { Icon } from '@talixo/icon'
import { TextInput } from '@talixo/text-input'

export const moduleName = 'time-picker'

const A_KEY = 65
const P_KEY = 80

const propTypes = {
  /** Additional class name. */
  className: PropTypes.string,

  /** Event called after input has lost focus. */
  onBlur: PropTypes.func,

  /** Format of time. */
  format: PropTypes.oneOf(['HH', 'hh A', 'mm']).isRequired,

  /** Time object. */
  value: PropTypes.object.isRequired
}

const defaultProps = {
}

const HOURS_24 = 'HH'
const HOURS_12 = 'hh A'
const MINUTES = 'mm'

/**
 * Component which represents time input
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {function} [props.onBlur]
 * @property {string} [props.format]
 * @property {object} [props.value]
 *
 * @property {object} state
 * @property {boolean} [state.open]
 * @property {string} [state.inputValue]
 * @property {string|null} [state.suffix]
 *
 * @class
 */
class TimeInput extends React.Component {
  state = {
    open: false,
    inputValue: '',
    suffix: null
  }

  componentDidMount () {
    this.formatValue(this.props)
  }

  componentWillReceiveProps (nextProps) {
    if (this.props.value !== nextProps.value || this.props.format !== nextProps.format) {
      this.formatValue(nextProps)
    }
  }

  /**
   * Formats value to desired format.
   *
   * @param {object} props
   * @returns {React.Element}
   */
  formatValue = (props) => {
    const { format, value } = props

    // Format value according to provided format
    const inputValue = format === HOURS_12
      ? moment(value).format('h')
      : moment(value).format(format)

    // Format suffix if needed
    let suffix = format !== HOURS_12
      ? null
      : moment(value).format('A')

    this.setState({ inputValue, suffix })
  }

  /**
   * Build control arrow.
   *
   * @returns {React.Element}
   */
  buildControl = () => {
    const { open } = this.state

    // Build class name for arrow
    const arrowClsName = buildClassName([ moduleName, 'input' ], null, 'arrow')

    return (
      <span className={arrowClsName}>
        <Icon name={open ? 'expand_less' : 'expand_more'} />
      </span>
    )
  }

  /**
   * Handles input value change.
   *
   * @param {string} value
   */
  handleChange = (value) => {
    const { format } = this.props

    // Parse value
    const parsedValue = isNaN(parseInt(value))
      ? 0
      : parseInt(value)

    let inputValue

    // Format value within desired range
    switch (format) {
      case HOURS_24:
        inputValue = Math.min(23, Math.max(0, parsedValue))
        break
      case HOURS_12:
        inputValue = Math.min(12, Math.max(0, parsedValue))
        break
      case MINUTES:
        inputValue = Math.min(59, Math.max(0, parsedValue))
        break
      default:
        inputValue = value
    }

    inputValue = inputValue === 0
      ? ''
      : inputValue.toString()

    this.setState({ inputValue })
  }

  /**
   * Handles input blur.
   */
  handleBlur = () => {
    const { onBlur } = this.props
    const { inputValue, suffix } = this.state

    if (onBlur) {
      onBlur(inputValue, suffix)
    }

    this.setState({ open: false })
  }

  /**
   * Handles input focus.
   */
  handleFocus = () => {
    this.setState({ open: true })
  }

  /**
   * Handle keyboard events on input.
   *
   * @param {SyntheticEvent|Event} event
   */
  handleInputKeyDown = (event) => {
    const { format } = this.props
    // If format is differnet than 'hh A' - return
    if (format !== HOURS_12) { return }

    // Prevent propagation if 'a' key is pressed and change suffix to 'AM'
    if (event.which === A_KEY) {
      event.stopPropagation()
      let suffix = 'AM'
      this.setState({ suffix })
    }

    // Prevent propagation if 'p' key is pressed and change suffix to 'PM'
    if (event.which === P_KEY) {
      event.stopPropagation()
      let suffix = 'PM'
      this.setState({ suffix })
    }
  }

  /**
   * Render input.
   *
   * @returns {React.Element}
   */
  renderInput = () => {
    const { inputValue, open, suffix } = this.state

    // Build class name for input
    const inputClsName = buildClassName([ moduleName, 'input' ], null, { open })

    return (
      <TextInput
        className={inputClsName}
        onChange={this.handleChange}
        onBlur={this.handleBlur}
        onFocus={this.handleFocus}
        onKeyDown={this.handleInputKeyDown}
        right={this.buildControl()}
        type='number'
        suffix={suffix}
        value={inputValue}
        autoComplete='off'
      />
    )
  }

  render () {
    const { className, children, format, onBlur, value, ...passedProps } = this.props
    const { open } = this.state

    // Build class name for wrapper
    const wrapperClsName = buildClassName([ moduleName, 'input' ], className)

    return (
      <div className={wrapperClsName} {...passedProps}>
        {this.renderInput()}
        {open && children}
      </div>
    )
  }
}

TimeInput.propTypes = propTypes

TimeInput.defaultProps = defaultProps

export default TimeInput
