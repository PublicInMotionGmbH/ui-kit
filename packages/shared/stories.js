import React from 'react'
import { createStoriesFactory } from './story'

// Load first paragraph from README file
const readme = ''

// Create factories for story
const addStory = createStoriesFactory('Typography and basics', module)

// Stories

addStory('Headers', readme, () => (
  <div>
    <h1>h1. Header</h1>
    <h2>h2. Header</h2>
    <h3>h3. Header</h3>
    <h4>h4. Header</h4>
    <h5>h5. Header</h5>
    <h6>h6. Header</h6>

    <h1>h1. Header <small>Secondary text</small></h1>
    <h2>h2. Header <small>Secondary text</small></h2>
    <h3>h3. Header <small>Secondary text</small></h3>
    <h4>h4. Header <small>Secondary text</small></h4>
    <h5>h5. Header <small>Secondary text</small></h5>
    <h6>h6. Header <small>Secondary text</small></h6>
  </div>
))

addStory('Text styling', readme, () => (
  <div>
    <h2>Paragraphs</h2>

    <p>
      Lorem ipsum dolor sit amet enim. Etiam ullamcorper. Suspendisse a pellentesque dui, non felis.
      Maecenas malesuada elit lectus felis, malesuada ultricies.
      Curabitur et ligula. Ut molestie a, ultricies porta urna.
      Vestibulum commodo volutpat a, convallis ac, laoreet enim. Phasellus fermentum in, dolor.
    </p>

    <p>
      Pellentesque facilisis. Nulla imperdiet sit amet magna.
      Vestibulum dapibus, mauris nec malesuada fames ac turpis velit, rhoncus eu, luctus et interdum adipiscing wisi.
    </p>

    <p>
      Aliquam erat ac ipsum. Integer aliquam purus.
      Quisque lorem tortor fringilla sed, vestibulum id, eleifend justo vel bibendum sapien massa ac turpis
      faucibus orci luctus non, consectetuer lobortis quis, varius in, purus.
      Integer ultrices posuere cubilia Curae, Nulla ipsum dolor lacus, suscipit adipiscing.
      Cum sociis natoque penatibus et ultrices volutpat.
    </p>

    <h2>Inline styling</h2>

    <p>
      We may have some <strong>strong elements</strong> inside our text, as well as some <em>emphasises</em>.
      When we want to calculate 2<sup>3</sup> + x<sub>1</sub>, we may have problems, but it's easy to show it.
      <br />
      In some circumstances, mostly describing our API or other stuff we may also try to show some <code>short code snippet</code>.
      It happens that we may also use <kbd>Ctrl+C</kbd> to hint user with keyboard shortcuts.
    </p>

    <p>
      It may happen that we would like to <mark>highlight</mark> some content inside our dashboards, in case of some search engine.
    </p>

    <p>
      <del>We do not make any mistakes.</del> When we do, we can write it semantically marking that anyway.
    </p>

    <p>
      When we use some shortcuts like <abbr title='Looks good to me'>LGTM</abbr>, we can show such information for user.
    </p>

    <h2>Quotes and testimonials</h2>

    <blockquote>
      This was one of the best experiences I have ever had with a cab company.
      I had problems at the airport and the driver stayed with me for over an hour and helped me sort everything out.
      I would recommend this company to anyone. Thank you for such fabulous service!
    </blockquote>

    <blockquote>
      This was one of the best experiences I have ever had with a cab company.
      I had problems at the airport and the driver stayed with me for over an hour and helped me sort everything out.
      I would recommend this company to anyone. Thank you for such fabulous service!

      <footer>
        <cite>R.M.</cite>
      </footer>
    </blockquote>
  </div>
))

addStory('RTL: Headers', readme, () => (
  <div dir='rtl'>
    <h1>h1. כּוֹתֶרֶת</h1>
    <h2>h2. כּוֹתֶרֶת</h2>
    <h3>h3. כּוֹתֶרֶת</h3>
    <h4>h4. כּוֹתֶרֶת</h4>
    <h5>h5. כּוֹתֶרֶת</h5>
    <h6>h6. כּוֹתֶרֶת</h6>

    <h1>h1. כּוֹתֶרֶת <small>טקסט משני</small></h1>
    <h2>h2. כּוֹתֶרֶת <small>טקסט משני</small></h2>
    <h3>h3. כּוֹתֶרֶת <small>טקסט משני</small></h3>
    <h4>h4. כּוֹתֶרֶת <small>טקסט משני</small></h4>
    <h5>h5. כּוֹתֶרֶת <small>טקסט משני</small></h5>
    <h6>h6. כּוֹתֶרֶת <small>טקסט משני</small></h6>
  </div>
))

addStory('RTL: Text styling', readme, () => (
  <div dir='rtl'>
    <h2>פסקאות</h2>

    <p>
      אחד אל בדפים חרטומים, או פיסול צרפתית ויקיפדיה היא, אם מדריכים הגולשות כתב. ב עוד המחשב ברוכים עקרונות. ספינות מדריכים אקטואליה אחר על.
      זאת אם ביולוגיה אדריכלות, עוד ראשי וקשקש מתמטיקה ב. ויש בה רביעי תרומה לציין. זקוק המדינה גם עוד.
      זאת דת לתרום הבאים נוסחאות, אם פיסיקה סטטיסטיקה אווירונאוטיקה רבה.
    </p>

    <p>
      נפלו למנוע מונחונים או כדי. אחרות תקשורת בה אחר. מיזם יסוד רביעי בה מלא, הבקשה בדפים על זכר. כלל או העזרה ממונרכיה, כלל גם הטבע אנציקלופדיה, אנא אל שמות יסוד.
    </p>

    <p>
תיקונים והגולשים אל בדף, של בדף בישול יוצרים העברית. עזה אם והוא הארץ בידור, תחבורה משופרות עזה אם.
      מה רוסית בארגז הגולשות צ'ט, קבלו מוסיקה ב אחד, מדעי אדריכלות בדף בה. מדע שתפו ויקימדיה אם.
      אדריכלות אווירונאוטיקה אתה של, תרבות ולחבר צ'ט דת.
      לחשבון לויקיפדיה עזה אל, תנך דרכה בכפוף מיוחדים של.
      שתי מה רפואה אודות, כיצד רקטות סטטיסטיקה דת שמו. עוד אקראי וקשקש מיוחדים אם,
      לוח גם העזרה בהתייחסות. מה מונחים אספרנטו קרן. אינו לעריכה של חפש, בה מאמר מוגש אנא.
      בקר הרוח מתמטיקה את, בה החלה יסוד הגולשות קרן. בה לימודים ממונרכיה ואלקטרוניקה עזה,
      חופשית מועמדים מדע של. מדע עזרה אחרות המלצת של, דת יוני אנגלית כתב.
    </p>

    <h2>עיצוב מוטבע</h2>

    <p>
      אולי יש לנו<strong> כמה אלמנטים חזקים בתוך הטקסט </strong>שלנו, כמו גם כמה מדגיש.
      ב במקרים מסוימים, בעיקר לתאר את ה- API שלנו או דברים אחרים אנו עשויים<code> גם לנסות להציג קטע קוד קצר. זה קורה</code>כי אנו יכולים גם >להשתמ
      <kbd>Ctrl+C</kbd> לרמז למשתמש עם קיצורי מקשים.
    </p>

    <p>
      זה יכול לקרות כי אנחנו רוצים להדגיש קצת תוכן בתוך לוחות.<mark> המחוונים שלנו, במקרה של מנוע </mark>החיפוש
    </p>

    <p>
      <del>אנחנו לא עושים שום טעויו</del>ת. כאשר אנו עושים, אנחנו יכולים לכתוב את זה סמנטית סימון כי בכל מקרה.
    </p>

    <p>
      כאשר אנו משתמשים בקיצורי דרך מסוימים כמו <abbr title='נראה לי טוב'>LGTM</abbr>, אנו יכולים להציג מידע כזה עבור המשתמש.
    </p>

    <h2>ציטוטים ועדויות</h2>

    <blockquote>
      זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
    </blockquote>

    <blockquote>
      זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
      <footer>
        <cite>מה-שמו</cite>
      </footer>
    </blockquote>

    <div style={{ display: 'flex' }}>
      <blockquote style={{ marginLeft: '15px' }}>
        זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
  היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
  אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
        <footer>
          <cite>מה-שמו</cite>
        </footer>
      </blockquote>

      <blockquote>
        זה היה אחד החוויות הטובות ביותר שהיו לי אי פעם עם חברת מוניות.
  היו לי בעיות בשדה התעופה והנהג נשאר איתי יותר משעה ועזר לי למיין הכול.
  אני ממליץ על חברה זו לכל אחד. תודה לך על שירות נהדר
        <footer>
          <cite>מה-שמו</cite>
        </footer>
      </blockquote>
    </div>
  </div>
))
