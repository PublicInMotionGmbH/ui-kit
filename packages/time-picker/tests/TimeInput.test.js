import React from 'react'
import { shallow } from 'enzyme'

import { prefix } from '@talixo/shared'

import TimeInput, { moduleName } from '../src/TimeInput'

const clsName = prefix(moduleName)

const createWrapper = (props) => shallow(<TimeInput {...props} />)

jest.mock('moment', () => {
  const moment = require.requireActual('moment')
  return moment.utc
})

const value = new Date('2018-05-10T00:00:00')

describe('<TimeInput />', () => {
  it('renders correctly', () => {
    const format = 'HH'
    const wrapper = createWrapper({ format, value })

    expect(wrapper).toMatchSnapshot()
  })

  it('renders open correctly', () => {
    const format = 'HH'
    const wrapper = createWrapper({ format, value })
    wrapper.setState({ open: true })
    const input = wrapper.find('TextInput')

    expect(input.hasClass(`${clsName}__input--open`)).toEqual(true)
  })
})

describe('onBlur', () => {
  it('is called when input looses focus', () => {
    const format = 'HH'
    const onBlur = jest.fn()
    const wrapper = createWrapper({ format, onBlur, value })

    const input = wrapper.find('TextInput')
    input.simulate('change', { target: 2 })
    input.simulate('blur')
    expect(onBlur).toHaveBeenCalledTimes(1)
  })

  it('passes value correctly', () => {
    let selectedSuffix, selectedValue
    const format = 'HH'
    const onBlur = jest.fn().mockImplementation((inputValue, suffix) => {
      selectedValue = inputValue
      selectedSuffix = suffix
    })
    const wrapper = createWrapper({ format, onBlur, value })

    wrapper.setState({ inputValue: '22', suffix: 'AM' })
    const input = wrapper.find('TextInput')
    input.simulate('blur')
    expect(selectedValue).toEqual('22')
    expect(selectedSuffix).toEqual('AM')
  })

  it('closes menu', () => {
    const format = 'HH'
    const wrapper = createWrapper({ format, value })

    wrapper.setState({ open: true })
    const input = wrapper.find('TextInput')
    input.simulate('blur')

    expect(wrapper.state().open).toEqual(false)
  })
})

describe('onFocus', () => {
  it('is called when input gains focus', () => {
    const format = 'HH'
    const wrapper = createWrapper({ format, value })

    const input = wrapper.find('TextInput')
    input.simulate('focus')
    expect(wrapper.state().open).toEqual(true)
  })
})

describe('onKeyDown', () => {
  it('is ignored if format is not `hh A`', () => {
    const format = 'HH'
    const wrapper = createWrapper({ format, value })

    wrapper.setState({ suffix: 'PM' })
    const input = wrapper.find('TextInput')
    input.simulate('keyDown', {
      which: 65,
      stopPropagation: () => {}
    })

    expect(wrapper.state().suffix).toEqual('PM')
  })

  it('changes state.suffix to `AM` when `a` is pressed', () => {
    const format = 'hh A'
    const wrapper = createWrapper({ format, value })

    wrapper.setState({ suffix: 'PM' })
    const input = wrapper.find('TextInput')
    input.simulate('keyDown', {
      which: 65,
      stopPropagation: () => {}
    })

    expect(wrapper.state().suffix).toEqual('AM')
  })

  it('changes state.suffix to `PM` when `p` is pressed', () => {
    const format = 'hh A'
    const wrapper = createWrapper({ format, value })

    wrapper.setState({ suffix: 'AM' })
    const input = wrapper.find('TextInput')
    input.simulate('keyDown', {
      which: 80,
      stopPropagation: () => {}
    })

    expect(wrapper.state().suffix).toEqual('PM')
  })
})
