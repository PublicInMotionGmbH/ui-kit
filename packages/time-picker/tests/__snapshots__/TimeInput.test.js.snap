// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimeInput /> renders correctly 1`] = `
<div
  className="talixo-time-picker__input"
  disabled={false}
  readOnly={false}
>
  <TextInput
    InputComponent="input"
    autoComplete="off"
    className="talixo-time-picker__input"
    disabled={false}
    endCaretOnFocus={false}
    error={false}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    readOnly={false}
    right={
      <span
        className="talixo-time-picker__input talixo-time-picker__input--arrow"
      >
        <Icon
          name="expand_more"
        />
      </span>
    }
    suffix={null}
    type="number"
    value="00"
  />
</div>
`;
